### PR TITLE
MAINT/ENH: refactor np.random.x to use np.random.Generator

### LINKED ISSUE #5342: ENH/DOC: remove references to np.random.RandomState in favour of np.random.Generator
### Description

Once NumPy minimal version is set to 1.17 #5340, I propose to do a PR to remove from the documentation/code any reference to `RandomState` and wrong (not best practice) usage of random number generators.

We went through this in SciPy (doc part for now). More on the PR https://github.com/scipy/scipy/pull/13778

### Additional context

Since NumPy 1.17, the [canonical way](https://numpy.org/doc/stable/reference/random/index.html) to create random numbers is:

```python
import numpy as np
from numpy.random import default_rng
rng = default_rng()
```
And If a seed is to be used. User should not use a common value such as 0 or 123456. This is to prevent bias (everyone is using the same values) and ensure sufficient entropy for the generator. More [details](https://numpy.org/doc/stable/reference/random/bit_generators/index.html#seeding-and-entropy). The good practice (from NumPy's doc) is to generate a seed like this:

```python
from numpy.random import SeedSequence
print(SeedSequence().entropy)
```

Then you can use the value like
```python
rng = default_rng(my_seed_generated_and_not_common_value)
```

### PR DESCRIPTION
**Pull request by tupui**
## Description

Closes #5342

Remove usage of `np.random.RandomState` in code and examples in favour of `np.random.Generator`. Only kept in tests as this is still the recommended way.
Also this change all direct calls to `np.random.x` to use a `np.random.Generator` (also recommended).

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by tupui** [comment]
@jni I started to work on this. I did a first pass at everything. I pushed so that this can get feedbacks and also to see the CI.

File: skimage/util/noise.py
Comment (grlee77) [inline skimage/util/noise.py:214]: _bernoulli expects `rng` rather than seed? I think that is causing some of the test failures

**Comment by grlee77** [comment]
I see that in a couple of places a new argument was added:
1.) added `seed` kwarg to `medial_axis`
2.) added `random_seed` kwarg to `unsupervised_wiener`
I think these should probably be keyword-only arguments (i.e. `,*, seed=None`)

it seems we were not very consistent in naming the argument `seed` vs. `random_seed` across the library (I think I also saw `sample_seed` in one place). Making this more consistent is something to consider separately for the 1.0 API.

File: skimage/feature/brief.py
Comment (grlee77) [inline skimage/feature/brief.py:121]: why change the default here?

Comment (grlee77) [reply skimage/feature/brief.py:121]: If we leave it `sample_seed=1` I think this would give different random values than with the legacy `RandomState(1)`, but it would at least still be the same values each time it is run?

**Comment by tupui** [comment]
> I see that in a couple of places a new argument was added:
> 
> 1.) added `seed` kwarg to `medial_axis`
> 
> 2.) added `random_seed` kwarg to `unsupervised_wiener`
> 
> I think these should probably be keyword-only arguments (i.e. `,*, seed=None`)
> 
> 
> 
> it seems we were not very consistent in naming the argument `seed` vs. `random_seed` across the library (I think I also saw `sample_seed` in one place). Making this more consistent is something to consider separately for the 1.0 API.
> 
> 

Oh yes sorry i forgot to mention that. I can make them keyword only. Let me know the name of the keyword I should use.

File: skimage/feature/brief.py
Comment (tupui) [reply skimage/feature/brief.py:121]: I would leave the decision to you here as I don't have any background here 😅. I just checked the doctests we're passing here, which is the case. In general it's not a good idea to have seeds set in stones but here it might be a special case.

Comment (grlee77) [reply skimage/feature/brief.py:121]: Let's see what others think. I have never used the feature detection module much myself.

**Comment by jni** [comment]
@tupui Gah, looks like something went wrong with a rebase...?

**Comment by tupui** [comment]
Yes, will fix it. I had master which was in front of the changes I wanted to get from the NumPy bump PR.

File: skimage/_shared/tests/test_utils.py
Comment (jni) [inline skimage/_shared/tests/test_utils.py]: Please revert this...

File: skimage/feature/brief.py
Comment (jni) [inline skimage/feature/brief.py:8]: Revert these changes, they should be relative imports.

File: skimage/feature/brief.py
Comment (jni) [reply skimage/feature/brief.py:121]: Yeah tbh this would be a breaking change, I agree that it should not have been 1 to begin with but now we are sort of stuck with it. I'd revert it, and leave the decision of what to do with it to a later issue.

File: skimage/measure/fit.py
Comment (jni) [inline skimage/measure/fit.py:740]: Can we perhaps continue setting the seed here (and below) so that the doctest can still be run? In some cases it's unclear whether things are suitably tested without the doctest...!

File: skimage/util/noise.py
Comment (jni) [inline skimage/util/noise.py:8]:
```suggestion
def _bernoulli(p, shape, *, random_state):
```

**Review by jni — COMMENTED** [review]
Thanks @tupui! I've left some minor comments but generally 👍 on the PR. Should be easy to get this over the line. 🙏

File: doc/examples/applications/plot_rank_filters.py
Comment (tupui) [reply doc/examples/applications/plot_rank_filters.py:70]: Haha totally, I also still need to check NumPy's doc from time to time because I am using the wrong methods.

File: skimage/feature/brief.py
Comment (tupui) [reply skimage/feature/brief.py]: I reverted to 1 then.

File: skimage/measure/fit.py
Comment (tupui) [reply skimage/measure/fit.py:740]: Do you have a CI job for the doctests?

We had some discussions for SciPy (and still are having some on Astropy) and we came up with a monkeypatching of the line `>>> rng = np.random.default_rng()` to include a seed. To me this is no practical but it works. Here it does not seems that you have much examples which would really need a seed. Hence I would advice not to use any. And if there are things that are tested in the examples which are not in the tests, then we should simply move these to tests. In the end it's double the work for you maintainers to maintain 2 test suites otherwise.

Comment (jni) [reply skimage/measure/fit.py:740]: Yes, the doctests are configured run by default. imho the double effort is in the other direction — if we add tests in addition to doctests (which no longer become doctests), then we have to maintain them in two places.

Comment (tupui) [reply skimage/measure/fit.py:740]: I was more thinking about moving than duplicating. IMO doctests should only serve to check that examples are still aligned with the implementation.

Anyway, what I can propose is to either put a seed when really necessary and comment with a `DO NOT COPY`. So that you could still use them for testing. For the seed itself, I would not use the same for all examples and I would get it using `np.random.SeedSequence`. Would that be ok?

Comment (jni) [reply skimage/measure/fit.py:740]: Yup, that all sounds fine! Sometimes an example usage is close enough to what a test would be that you can kill two birds with one stone — and we do this liberally at skimage. :blush: :grimacing:

**Comment by tupui** [comment]
I fixed the tests. I just have the following issues. For the first 2, I guess I need to update the files? For the last 2 I am not sure why this crashes.

Can you help with that @jni?

I will have a look at the examples now.

```bash
_______________________________ test_3d_vs_fiji ________________________________

        img_f = io.imread(fetch("data/_blobs_3d_fiji_skeleton.tif"))

___________________________ test_unsupervised_wiener ___________________________
    
        path = fetch('restoration/tests/camera_unsup.npy')

________________________________ test_localvar _________________________________

passing but no stable with different seeds

_________________________________ test_speckle _________________________________

scikit-image/skimage/util/tests/test_random_noise.py:143: AssertionError

_________________________________ test_poisson _________________________________

scikit-image/skimage/util/tests/test_random_noise.py:154: AssertionError
```

**Comment by grlee77** [comment]
> Can you help with that @jni?

I fixed the test_speckle and test_poisson and regenerated the data for `test_unsupervised_wiener`. For `test_3d_vs_fiji`, I assume the original reference data was generated using Fiji? I didn't try updating that one here.

**Comment by grlee77** [comment]
I think the only failing tests now are the ones related to the change in the sha256 for the 3 files that were updated. In some of the CI logs these are being downloaded from the `main` branch rather than this one, which is causing the mismatch:

for example:
```
Downloading file 'data/_blobs_3d_fiji_skeleton.tif' from 'https://github.com/scikit-image/scikit-image/raw/main/skimage/data/_blobs_3d_fiji_skeleton.tif' to '/home/runner/.cache/scikit-image/main'
```
so it is downloading the old version of the file, but comparing it to the sha256 for the new one

I assume this will just work once we merge this branch.

**Comment by tupui** [comment]
Thanks you @grlee77! This is great. Is there anything else I should do?

File: skimage/data/_binary_blobs.py
Comment (hmaarrfk) [inline skimage/data/_binary_blobs.py:36]: Is this really the better way?

**Comment by hmaarrfk** [comment]
i'm curious how default_rng will change between versions of numpy. Is it guaranteed to be the same, or will it vary as things evolve.

**Comment by tupui** [comment]
> i'm curious how default_rng will change between versions of numpy. Is it guaranteed to be the same, or will it vary as things evolve.

`np.random.RandomState` is guaranteed to be the same but has known issues and is only recommended for testing purposes. See [NEP19](https://numpy.org/neps/nep-0019-rng-policy.html). It is unlikely that `default_rng` will change often, but they keep it open so that they could fix bugs or improve the generators. Things that got intractable with the old architecture.

File: skimage/data/_binary_blobs.py
Comment (tupui) [reply skimage/data/_binary_blobs.py:36]: In this case I could use a `seed` generated randomly and add `# do not copy` as done in other examples (this is not perfect but we agreed in previous comments that this was an OK compromise). I guess I forgot this one. I can do this if preferred.

Comment (hmaarrfk) [reply skimage/data/_binary_blobs.py:36]: hmm ok. yeah i would rather people not copy the seed. I see this kind of copy from examples stuff too often.

File: skimage/data/_binary_blobs.py
Comment (hmaarrfk) [inline skimage/data/_binary_blobs.py:21]: This is a new feature and should be highlighted. I Think we things like `.. versionadded::`

Comment (tupui) [reply skimage/data/_binary_blobs.py:21]: Indeed, also not clear what to choose as per #5359. And probably I should add it as keyword argument only with `..., *, seed`. What should I put for the version?

File: skimage/measure/fit.py
Comment (tupui) [reply skimage/measure/fit.py]: @hmaarrfk we had this discussion. So in the example you noted, I missed to do this. Let me know what you choose so I modify it accordingly if needed.

File: skimage/data/_binary_blobs.py
Comment (grlee77) [reply skimage/data/_binary_blobs.py:21]: Yes, please make any added parameters keyword-only. 

Use `0.19` for the versionadded statement.

Comment (tupui) [reply skimage/data/_binary_blobs.py:21]: ok thanks @grlee77. And for the naming?

**Comment by grlee77** [comment]
I think the only other task aside from the versionadded and keyword-only stuff mentioned above is adding something about this to `doc/release/release_dev.rst`. The new keywords can be mentioned under the "New Features" section. We should probably also mention that seeds will not give identical results to prior releases due to the change in the underlying generator used.

To help avoid conflicts you may need to merge master to this branch first since a lot of entries left over from the 0.18 release were recently removed from that file in #5367.

File: skimage/data/_binary_blobs.py
Comment (grlee77) [reply skimage/data/_binary_blobs.py:21]: I don't think we reached any conclusion in #5359 on the preferred name, but I am personally leaning toward `random_state`. Unless there are objections, we could use that for any newly added parameter here. I wouldn't rename any of the pre-existing parameters in this PR. Let's leave that to a potential follow-up.

Comment (tupui) [reply skimage/data/_binary_blobs.py:21]: To be clear, in this case I did not add the parameter but just the documentation and called behind `default_rng ` instead of `RandomState`. But there are other places where I added the parameter. Correct me if I am wrong, but I should just add versionadded info to the new places but not here.

**Comment by tupui** [comment]
@grlee77 I think this is ready on my side. I addressed the points you listed and merged master.

**Review by grlee77 — APPROVED** [review]
Thanks @tupui, it looks good to me.

**Comment by grlee77** [comment]
@hmaarrfk or @jni, can one of you take a quick look at this one again and merge if it seems okay to you? 

The CI failures here look like they are because I updated a few of the test files and pooch is using the version from main instead of the ones from this branch.

File: doc/release/release_dev.rst
Comment (mkcor) [inline doc/release/release_dev.rst:18]:
```suggestion
- Add a new keyword-only parameter ``random_state`` to
```

File: skimage/data/_binary_blobs.py
Comment (mkcor) [inline skimage/data/_binary_blobs.py:22]:
```suggestion
        If `seed` is None, the `numpy.random.Generator` singleton is used.
```

File: skimage/draw/_random_shapes.py
Comment (mkcor) [inline skimage/draw/_random_shapes.py:356]:
```suggestion
        If `random_seed` is None, the `numpy.random.Generator` singleton is
```

File: skimage/feature/brief.py
Comment (mkcor) [inline skimage/feature/brief.py:37]:
```suggestion
        If `sample_seed` is None, the `numpy.random.Generator` singleton is
```

File: skimage/feature/haar.py
Comment (mkcor) [inline skimage/feature/haar.py:260]:
```suggestion
        If `random_state` is None, the `numpy.random.Generator` singleton is
```

File: skimage/future/graph/graph_cut.py
Comment (mkcor) [inline skimage/future/graph/graph_cut.py:103]:
```suggestion
        If `random_state` is None, the `numpy.random.Generator` singleton is
```

File: skimage/measure/fit.py
Comment (mkcor) [inline skimage/measure/fit.py:707]:
```suggestion
        If `random_state` is None, the `numpy.random.Generator` singleton is
```

File: skimage/morphology/_skeletonize.py
Comment (mkcor) [inline skimage/morphology/_skeletonize.py:374]:
```suggestion
        If `random_state` is None, the `numpy.random.Generator` singleton is
```

File: skimage/morphology/tests/test_skeletonize_3d.py
Comment (mkcor) [inline skimage/morphology/tests/test_skeletonize_3d.py:178]:
```suggestion
    # the skeleton generated by FIJI (Plugins > Skeleton > Skeletonize)
```
or "(Plugins>Skeleton>Skeletonize)" or "(Plugins->Skeleton->Skeletonize)"

File: skimage/restoration/deconvolution.py
Comment (mkcor) [inline skimage/restoration/deconvolution.py:167]:
```suggestion
        If `random_state` is None, the `numpy.random.Generator` singleton is
```

File: skimage/segmentation/_quickshift_cy.pyx
Comment (mkcor) [inline skimage/segmentation/_quickshift_cy.pyx:37]:
```suggestion
        If `random_seed` is None, the `numpy.random.Generator` singleton
```

File: skimage/transform/_hough_transform.pyx
Comment (mkcor) [inline skimage/transform/_hough_transform.pyx:351]:
```suggestion
        If `seed` is None, the `numpy.random.Generator` singleton is used.
```

File: skimage/util/noise.py
Comment (mkcor) [inline skimage/util/noise.py:63]:
```suggestion
        If `seed` is None, the `numpy.random.Generator` singleton is
```

**Review by mkcor — COMMENTED** [review]
Thank you, @tupui! Would you mind taking care of the (now) conflicting files? Sorry this has been dragging on...

**Comment by grlee77** [comment]
The conflicts are related to other float32 PRs of mine that were merged, so I can help resolve them here.

**Comment by grlee77** [comment]
One issue that came up in resolving conflicts is related to the following lines in unsupervised_wiener:

```Python
    rng = np.random.default_rng(random_state)
```
```Python
        _rand1 = rng.standard_normal(data_spectrum.shape)
        _rand1 = _rand1.astype(float_type, copy=False)
        _rand2 = rng.standard_normal(data_spectrum.shape)
        _rand2 = _rand2.astype(float_type, copy=False)
        excursion = np.sqrt(0.5 / precision) * (_rand1 + 1j * _rand2)
```

Here it should be more efficient to pass a `dtype` argument to `standard_normal`. 
```Python
        _rand1 = rng.standard_normal(data_spectrum.shape, dtype=float_type)
        _rand2 = rng.standard_normal(data_spectrum.shape, dtype=float_type)
        excursion = np.sqrt(0.5 / precision) * (_rand1 + 1j * _rand2)
```

However, this results in a different stream of random values for float32 vs. float64, which breaks comparison to the existing float64 reference data in `test_unsupervised_wiener`.

Here I used `astype` after generating float64 random values so that we could reuse the same reference data, but perhaps we should also store separate reference data for the float32 case instead?

**Comment by tupui** [comment]
Thanks @grlee77 for updating the PR.
Sorry I cannot help myself until end of June. I am away without a computer.

**Comment by mkcor** [comment]
> However, this results in a different stream of random values for float32 vs. float64, which breaks comparison to the existing float64 reference data in `test_unsupervised_wiener`.

@grlee77 ok.

> Here I used astype after generating float64 random values so that we could reuse the same reference data, but perhaps we should also store separate reference data for the float32 case instead?

Can we do this (in a subsequent PR) by parametrizing `test_unsupervised_wiener` with `[np.float32, np.float64]`?

**Comment by mkcor** [comment]
> Can we do this (in a subsequent PR) by parametrizing `test_unsupervised_wiener` with `[np.float32, np.float64]`?

Oh, it looks like `test_unsupervised_wiener` could be parametrized(?) with `[np.float16, np.float32, np.float64]` from (screenshot of failing "linux-cp3.7-optional-deps" tests):
linux_cp37_optnl_deps
? Although I'm not seeing this in the PR...

I understand that the first two tests  (`data/tests/test_data.py::test_download_all_with_pooch` and `morphology/tests/test_skeletonize_3d.py::test_3d_vs_fiji`) fail because of hash mismatches, but what about `test_unsupervised_wiener`?

**Comment by grlee77** [comment]
> I understand that the first two tests (data/tests/test_data.py::test_download_all_with_pooch and morphology/tests/test_skeletonize_3d.py::test_3d_vs_fiji) fail because of hash mismatches, but what about test_unsupervised_wiener?

This failure is for the same reason (the data files used in the tests had to be updated since the random values generated within `unsupervised_wiener` are different after this PR).

**Review by mkcor — APPROVED** [review]

**Comment by tupui** [comment]
Thanks everyone for your help! Great community 😃

**Pull request merged**
